### PR TITLE
Added missing trigonometric functions(sech , csch , acsch)

### PR DESCRIPTION
**Pull request by CodeMaxx**
#777 
@certik @isuruf @Sumith1896

### CODE REVIEW

File: symengine/derivative.cpp [hunk 1]
```diff
@@ -115,6 +115,12 @@ static RCP<const Basic> diff(const CLASS &self, \
                 self.get_arg()->diff(x));
     }
 
+    static RCP<const Basic> diff(const ACsch &self,
+            const RCP<const Symbol> &x) {
+        return mul(div(minus_one, mul(sqrt(add(one, pow(self.get_arg(), i2))),
+                abs(self.get_arg()))), self.get_arg()->diff(x));
```
Comment (isuruf) [inline symengine/derivative.cpp]: This is not true for complex arguments. Use `-1/(x^2*sqrt(1/x^2 + 1))`

**Comment by isuruf** [comment]
I think you forgot to implement the class `ACsch`  in `functions.h`

File: symengine/derivative.cpp (hunk 1 above)
Comment (CodeMaxx) [inline symengine/derivative.cpp]: Chaged.

File: symengine/functions.h [hunk 2]
```diff
@@ -713,6 +761,27 @@ class ASinh : public HyperbolicFunction {
...
+    /*! Equality comparator
+     * \param o - Object to be compared with
+     * \return whether the 2 objects are equal
+     * */
+    virtual bool __eq__(const Basic &o) const;
+    virtual int compare(const Basic &o) const;
+    //! \return `true` if canonical
+    bool is_canonical(const RCP<const Basic> &arg) const;
+    //! \return Canonicalized acsch
+    virtual RCP<const Basic> create(const RCP<const Basic> &arg) const;
```
Comment (isuruf) [inline symengine/functions.h]: Is this implemented?

File: symengine/functions.h [hunk 3]
```diff
@@ -734,6 +803,8 @@ class ACosh: public HyperbolicFunction {
 //! Canonicalize ACosh:
 RCP<const Basic> acosh(const RCP<const Basic> &arg);
 
+
+
```
Comment (isuruf) [inline symengine/functions.h]: Remove these unnecessary lines

File: symengine/real_double.cpp [hunk 4]
```diff
@@ -80,10 +80,18 @@ class EvaluateDouble : public Evaluate {
...
+        return number(std::csch(static_cast<const T &>(x).i));
+    }
     virtual RCP<const Basic> cosh(const Basic &x) const override {
         SYMENGINE_ASSERT(is_a<T>(x))
         return number(std::cosh(static_cast<const T &>(x).i));
     }
+    virtual RCP<const Basic> sech(const Basic &x) const override {
+        SYMENGINE_ASSERT(is_a<T>(x))
+        return number(std::sech(static_cast<const T &>(x).i));
+    }
```
Comment (isuruf) [inline symengine/real_double.cpp]: Can you add `acsch` as well?

File: symengine/real_mpfr.cpp [hunk 5]
```diff
@@ -737,12 +737,24 @@ class EvaluateMPFR : public Evaluate {
...
         mpfr_class t(static_cast<const RealMPFR &>(x).i.get_prec());
         mpfr_cosh(t.get_mpfr_t(), static_cast<const RealMPFR &>(x).i.get_mpfr_t(), MPFR_RNDN);
         return real_mpfr(std::move(t));
     }
+    virtual RCP<const Basic> sech(const Basic &x) const override {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_class t(static_cast<const RealMPFR &>(x).i.get_prec());
+        mpfr_sech(t.get_mpfr_t(), static_cast<const RealMPFR &>(x).i.get_mpfr_t(), MPFR_RNDN);
+        return real_mpfr(std::move(t));
+    }
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: Same here

File: symengine/functions.cpp [hunk 6]
```diff
@@ -1847,6 +1847,72 @@ RCP<const Basic>  Sinh::expand_as_exp() const
...
+bool Csch::__eq__(const Basic &o) const
+{
+    if (is_a<Csch>(o) and
+        eq(*get_arg(), *(static_cast<const Csch &>(o).get_arg())))
+        return true;
+    else
+        return false;
+}
+
+int Csch::compare(const Basic &o) constraints
```
Comment (isuruf) [inline symengine/functions.cpp]: `constraints` to `const`

File: symengine/functions.cpp [hunk 7]
```diff
@@ -1847,6 +1847,72 @@ RCP<const Basic>  Sinh::expand_as_exp() const
...
+    if (is_a_Number(*arg)) {
+        RCP<const Number> _arg = rcp_static_cast<const Number>(arg);
+        if (not _arg->is_exact()) {
+            return _arg->get_eval().csch(*_arg);
+        } else if (_arg->is_negative()) {
+            return csch(zero->sub(*_arg));
+        }
+    }
+    if (could_extract_minus(arg)) {
+        return csch(neg(arg));
```
Comment (isuruf) [inline symengine/functions.cpp]: This should be `neg(csch(neg(arg)))`

File: symengine/functions.cpp [hunk 8]
```diff
@@ -1847,6 +1847,72 @@ RCP<const Basic>  Sinh::expand_as_exp() const
...
+    if (eq(*arg, *zero)) {
+        // Answer is infinity. Yet to be implemented in SymEngine
+        throw std::runtime_error("Not implemented.");
+    }
+    if (is_a_Number(*arg)) {
+        RCP<const Number> _arg = rcp_static_cast<const Number>(arg);
+        if (not _arg->is_exact()) {
+            return _arg->get_eval().csch(*_arg);
+        } else if (_arg->is_negative()) {
+            return csch(zero->sub(*_arg));
```
Comment (isuruf) [inline symengine/functions.cpp]: Same here

**Comment by CodeMaxx** [comment]
@isuruf I'll fix this in by thursday. Have an exam.

File: symengine/functions.h (hunk 2 above)
Comment (CodeMaxx) [inline symengine/functions.h]: Done. Had missed this somehow.

File: symengine/complex_mpc.cpp [hunk 9]
```diff
@@ -689,6 +703,13 @@ class EvaluateMPC : public Evaluate {
         mpc_asinh(t.get_mpc_t(), static_cast<const ComplexMPC &>(x).i.get_mpc_t(), MPFR_RNDN);
         return complex_mpc(std::move(t));
     }
+    virtual RCP<const Basic> acsch(const Basic &x) const {
+        SYMENGINE_ASSERT(is_a<ComplexMPC>(x))
+        mpc_class t(static_cast<const ComplexMPC &>(x).i.get_prec());
+        mpc_ui_div(t.get_mpc_t(), 1, static_cast<const ComplexMPC &>(x).i.get_mpc_t(), MPFR_RNDN);
+        mpc_atanh(t.get_mpc_t(), t.get_mpc_t(), MPFR_RNDN);
```
Comment (isuruf) [inline symengine/complex_mpc.cpp]: Is this correct?

Comment (CodeMaxx) [inline symengine/complex_mpc.cpp]: No its not. Should be asinh.

File: symengine/real_mpfr.cpp [hunk 10]
```diff
@@ -761,6 +773,16 @@ class EvaluateMPFR : public Evaluate {
         mpfr_asinh(t.get_mpfr_t(), static_cast<const RealMPFR &>(x).i.get_mpfr_t(), MPFR_RNDN);
         return real_mpfr(std::move(t));
     }
+    virtual RCP<const Basic> acsch(const Basic &x) const override {
+        SYMENGINE_ASSERT(is_a<RealMPFR>(x))
+        mpfr_srcptr x_ = static_cast<const RealMPFR &>(x).i.get_mpfr_t();
+        if (mpfr_cmp_si(x_, 1) >= 0 or mpfr_cmp_si(x_, -1) <= 0) {
```
Comment (isuruf) [inline symengine/real_mpfr.cpp]: this check is unnecessary

**Comment by isuruf** [comment]
Looks good to me. Great job.
Thanks for being very thorough and implementing all of the functionality.

**Comment by certik** [comment]
Should we add at least some tests for the new classes?

**Comment by srajangarg** [comment]
Yes, you can add the same test cases as of the other (similar) functions.

**Comment by CodeMaxx** [comment]
I'll do it by tomorrow

**Comment by certik** [comment]
This looks good to me. @isuruf any objections?

**Comment by isuruf** [comment]
Looks good to me

**Pull request merged**